server: inline `conn.checkMaxConnections()`, fix max connection race condition

ecwall opened the pull request:
Informs #105448

Inlines `conn.checkMaxConnections()` into `conn.processCommandsAsync()` and fixes a race condition in `conn.checkMaxConnections()` between reading nonRootConnectionCount and updating connectionCount.

ecwall committed:
server: fix max connection race condition

Informs #105448

Fixes a race condition between reading nonRootConnectionCount and
updating connectionCount. This could have allowed more than the
maximum number of connections to be estabilished.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105450)
<!-- Reviewable:end -->

knz commented:
I think this PR is not well-organized.

There are 3 different changes here:
-  fixes a race condition in conn.checkMaxConnections()
- inlines checkMaxConnections
- inlines sendInitialConnData

At the very least they should be 3 different commits.
The first one should have a release note.
I have no comment on the first 2 changes; however I take issue with the 3rd one. Functions/methods should be short. The separation of `sendInitialConnData` is beneficial to the reading flow. If you find otherwise, this may be due to a lack of docstring on that function. Adding the missing doc seems like the better improvement.

ecwall committed:
server: inline conn.checkMaxConnections()

Informs #105448

This is only called in only location and does not have tests.

Release note: None

ecwall commented:
> I think this PR is not well-organized.
> 
> There are 3 different changes here:
> 
> * fixes a race condition in conn.checkMaxConnections()
> * inlines checkMaxConnections
> * inlines sendInitialConnData
> 
> At the very least they should be 3 different commits. The first one should have a release note. I have no comment on the first 2 changes; however I take issue with the 3rd one. Functions/methods should be short. The separation of `sendInitialConnData` is beneficial to the reading flow. If you find otherwise, this may be due to a lack of docstring on that function. Adding the missing doc seems like the better improvement.

I removed `inlines sendInitialConnData` from this PR and split the other commit into 2 commits.

rafiss reviewed (changes requested):
looks good, just had some small comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105450#-:-NZ7arOD6Cv_VF4I8Dpj:bw76bhu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @knz)

___
*[`pkg/sql/conn_executor.go` line 845 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105450#-NZ7_3Ih0XlbhilGCJGD:-NZ7_3Ih0XlbhilGCJGE:bfjp9ui) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052412fc4421222725b993ec752d4743b440878/pkg/sql/conn_executor.go#L845)):*
> ```Go
> 	maxNumNonRootConnectionsValue := maxNumNonRootConnections.Get(sv)
> 	maxNumConnectionsValue := maxNumNonAdminConnections.Get(sv)
> 	maxNumNonRootConnectionsReason := maxNumNonRootConnectionsReason.Get(sv)
> ```

nit: it would be nicer to not reuse the same name
___
*[`pkg/sql/conn_executor.go` line 882 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105450#-NZ7aL114qMjmeVyrhh7:-NZ7aL120GOQ2ALzhWbu:b-32hiq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052412fc4421222725b993ec752d4743b440878/pkg/sql/conn_executor.go#L882)):*
> ```Go
> 	}()
> 	if maxNumNonRootConnectionsExceeded {
> 		err = errors.WithHintf(
> ```

nit: i think it would be better to `return nil, err` here.

that way there's no risk of the caller incorrectly calling `decrementConnectionCount` if the `err` is non nil
___
*[`pkg/sql/conn_executor.go` line 883 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105450#-NZ7ZpEF8TbwrMCHeC1Y:-NZ7ZpEF8TbwrMCHeC1Z:bk90o7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052412fc4421222725b993ec752d4743b440878/pkg/sql/conn_executor.go#L883)):*
> ```Go
> 	if maxNumNonRootConnectionsExceeded {
> 		err = errors.WithHintf(
> 			pgerror.Newf(pgcode.TooManyConnections, "%s", maxNumNonRootConnectionsReason),
> ```

nit: this should be 
```
pgerror.Newf(pgcode.TooManyConnections, "%s", redact.SafeString(maxNumNonRootConnectionsReason)),
```

otherwise all the error text would get redacted in splunk

(this was a bug from before your PR)
___
*[`pkg/sql/conn_executor.go` line 888 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105450#-NZ7aXDS4rk0ORw0A5UI:-NZ7aXDS4rk0ORw0A5UJ:b-32hiq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052412fc4421222725b993ec752d4743b440878/pkg/sql/conn_executor.go#L888)):*
> ```Go
> 		)
> 	} else if maxNumConnectionsExceeded {
> 		err = errors.WithHintf(
> ```

nit: i think it would be better to `return nil, err` here.

that way there's no risk of the caller incorrectly calling `decrementConnectionCount` if the `err` is non nil
___
*[`pkg/sql/conn_executor.go` line 895 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105450#-NZ7aXiT0lXHGqIsUURP:-NZ7aXiT0lXHGqIsUURQ:b-o0grqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052412fc4421222725b993ec752d4743b440878/pkg/sql/conn_executor.go#L895)):*
> ```Go
> 		)
> 	}
> 	return
> ```

nit: i think it would be more explicit to make this `return decrementConnectionCount, nil` here. the implicit return values made this a bit hard to read


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105450#-:-NZD6pSf-4qiMl-EgOZ_:b-7nbn20)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @knz)



<!-- Sent from Reviewable.io -->

ecwall commented:
bors r=rafiss

ecwall commented:
bors r=rafiss

ecwall commented:
bors r=rafiss
